### PR TITLE
feat: voting history for fellowship

### LINKED ISSUE #2522: Fellowship. Voting history for referendum
Figure out where to get the referenda voting history. Options are:
1. Subsquare https://collectives.subsquare.io/
2. Polkassembly https://collectives.polkassembly.io/
3. Implement SubQuery indexer

### PR DESCRIPTION
**Pull request by Asmadek**
* add view voting history button for fellowship
* show modal with votes

### CODE REVIEW

File: src/renderer/domains/collectives/model/votes/mapper.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { type Vote } from './types';
+
+export const mapVote = (): Vote => {
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/votes/mapper.ts]: Why?

File: src/renderer/domains/collectives/model/votes/service.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { type Vote } from './types';
+
+const sortVotes = (a: Vote, b: Vote) => b.votes - a.votes;
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/votes/service.ts]: is it necessary? I think it doesn't deserve own service

File: src/renderer/features/fellowship-votes/index.ts
Comment (johnthecat) [inline src/renderer/features/fellowship-votes/index.ts:1]: I think feature itself should be called fellowship-voting-history

File: src/renderer/domains/collectives/index.ts [hunk 3]
```diff
@@ -27,3 +27,7 @@ export type {
 export type { Voting, VotingTransaction } from './model/voting/types';
 
 export type { Member } from './model/members/types';
+
+export type { Vote } from './model/votes/types';
+
+export { votesService } from './model/votes/service';
```
Comment (johnthecat) [inline src/renderer/domains/collectives/index.ts]: reexport it in `collectiveDomain`, `init.ts` file

File: src/renderer/domains/collectives/model/referendumMeta/model.ts [hunk 4]
```diff
@@ -64,11 +64,11 @@ const {
     return response;
   },
   map: (store, { params, result }) => {
-    const currectValue = pickNestedValue(store, params.palletType, params.chainId);
+    const currentValue = pickNestedValue(store, params.palletType, params.chainId);
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/referendumMeta/model.ts:67]: Tnx!!! 🔥

File: src/renderer/domains/collectives/model/votes/model.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  initial: {},
+  fn: async ({ api, palletType, referendumId }) => {
+    const votes = await collectivePallet.storage.votingEntries(palletType, api, referendumId);
+
+    return votes
+      .map(vote => {
+        if (!vote.vote) return;
+
+        return {
+          accountId: vote.key[1],
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/votes/model.ts]: such mappings should be hidden behind pallet adapter

File: src/renderer/features/fellowship-votes/components/Vote.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    store: membersModel.$members,
+    keys: [item.accountId],
+    fn: (members, [accountId]) => members[accountId] ?? null,
+  });
+
+  const address = toAddress(item.accountId, { prefix: chain.addressPrefix });
+
+  return (
+    <div className="flex items-center justify-between">
+      <div className="flex grow items-center gap-2 px-2 py-3 contain-inline-size">
```
Comment (johnthecat) [inline src/renderer/features/fellowship-votes/components/Vote.tsx]: It think, maybe we should move implementation from members feature as separated component to `ui-entities`? Smth like `RankedAccount`

File: src/renderer/shared/pallet/collective/storage.ts [hunk 7]
```diff
@@ -124,4 +124,18 @@ export const storage = {
       callback(votingResponseSchema.parse(response));
     });
   },
+
+  /**
+   * All votes on a given proposal.
+   */
+  votingEntries(type: PalletType, api: ApiPromise, referendum: ReferendumId) {
```
Comment (johnthecat) [inline src/renderer/shared/pallet/collective/storage.ts]: It's better to collapse it with `votings` method like it's done in [referenda](https://github.com/novasamatech/nova-spektr/blob/af89d11aebeaa978e708602808dcdc1b91fe5c34/src/renderer/shared/pallet/referenda/storage.ts#L45)

File: src/renderer/shared/pallet/collective/storage.ts [hunk 8]
```diff
@@ -124,4 +124,18 @@ export const storage = {
...
+    const schema = pjsSchema.vec(
+      pjsSchema.tupleMap(
+        ['key', pjsSchema.storageKey(pjsSchema.u32, pjsSchema.accountId)],
```
Comment (johnthecat) [inline src/renderer/shared/pallet/collective/storage.ts]: It's better to map tuple into record, like
```
[
  'key', 
  pjsSchema.storageKey(pjsSchema.u32, pjsSchema.accountId).map(([referendum, account]) => ({ referendum, account }))
]
```

Also, use `referendaPallet.schema.referendumId` instead of `pjsSchema.u32`

File: src/renderer/features/fellowship-votes/components/Vote.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+          </div>
+        </div>
+        <AccountExplorers accountId={item.accountId} chain={chain} />
+      </div>
+      <BodyText>{t('fellowship.votingHistory.votes', { count: item.votes })}</BodyText>
+    </div>
+  );
+};
+
+const Rank = ({ rank }: { rank: number }) => {
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/components/Vote.tsx]: I suggest to use pattern matching for readability
```javascript
const Rank = ({ rank }: { rank: number }) => {
  const rankVariants: Record<number, ComponentProps<typeof Label>['variant']> = {
    2: 'orange',
    3: 'red',
    4: 'purple',
    5: 'lightBlue',
    6: 'green',
    7: 'blue',
  };

  const variant = rankVariants[rank] || 'gray';

  return <Label variant={variant}>{rank.toString()}</Label>;
};
```

File: src/renderer/features/fellowship-votes/components/VotingHistory.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import { votesFeatureStatus } from '../model/status';
+import { votesModel } from '../model/votes';
+
+import { VotesModal } from './VotesModal';
+
+type Props = {
+  referendumId: ReferendumId | null;
+};
+
+export const VotingHistory = memo<Props>(({ referendumId }) => {
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/components/VotingHistory.tsx]: `memo` makes a difference here?

File: src/renderer/features/fellowship-votes/components/VotingHistory.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  useGate(votesFeatureStatus.gate);
+  useGate(votesModel.gate, { referendumId });
+  const { t } = useI18n();
+
+  const featureState = useUnit(votesFeatureStatus.state);
+  const pending = useUnit(votesModel.$pending);
+  const isNetworkDisabled = featureState.status === 'failed' && featureState.error.message === ERROR.networkDisabled;
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/components/VotingHistory.tsx]: Shouldn't it be calculated inside model?

File: src/renderer/features/fellowship-votes/components/VotingHistoryList.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import { Vote } from './Vote';
+import { VotingHistoryListEmptyState } from './VotingHistoryListEmptyState';
+
+type Props = {
+  items: VoteType[];
+  chain: Chain | null;
+  loading?: boolean;
+};
+
+export const VotingHistoryList = memo(({ items, chain, loading }: Props) => {
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/components/VotingHistoryList.tsx]: Why so many memos here and there? Every new `items` will start render because references for the old and new one won't be the same. Same situation for chain.
`memo` is fine for dumb components that accepts primitive data like string, number, bool and it fails whenever you pass arrays and objects.
Same point for the following `useMemo` we will end up with dozens of memoized data for each new `query`, `items` and `identity`
`useMemo` is fine for heavy operations where we're sure that we will match previously memoized data 
Speaking about heavy operations, why do they even happen inside UI component? Effector model is the best place for that 😉

File: src/renderer/features/fellowship-votes/model/members.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
+import { sample } from 'effector';
+import { or } from 'patronum';
+
+import { dictionary } from '@/shared/lib/utils';
+import { collectiveDomain } from '@/domains/collectives';
+
+import { fellowshipModel } from './fellowship';
+import { votesFeatureStatus } from './status';
+
+const $members = fellowshipModel.$store.map(x => dictionary(x?.members ?? [], 'accountId'));
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/model/members.ts]: ⚠️ I suggest to stop using `x` as a variable in production code ⚠️

File: src/renderer/features/fellowship-votes/model/status.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import { sample } from 'effector';
+
+import { createFeature } from '@/shared/effector';
+import { fellowshipNetworkFeature } from '@/features/fellowship-network';
+import { ERROR } from '../constants';
+
+export const votesFeatureStatus = createFeature({
+  name: 'votes',
+  input: fellowshipNetworkFeature.model.network.$network,
+  filter: input => {
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-votes/model/status.ts]: For readability:
```javascript
filter: input => {
    if (input.api.isConnected) return null;

    return {
          status: 'failed',
          type: 'warning',
          error: new Error(ERROR.networkDisabled),
        };
  },
```

**Review by tuul-wq — COMMENTED** [review]
I enjoy the structure!

File: src/renderer/features/fellowship-votes/components/Vote.tsx [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  active: boolean;
+};
+
+const Indicator = ({ active }: IndicatorProps) => {
+  return (
+    <div className="pointer-events-none relative flex h-4 w-4 items-center justify-center rounded-full bg-white">
+      <div
+        className={cnTw('h-2 w-2 rounded-full shadow-[0_0_0_1.5px]', {
+          ['bg-text-positive shadow-badge-green-background-default']: active,
+          ['bg-chip-text shadow-secondary-button-background']: !active,
```
Comment (sokolova-an) [inline src/renderer/features/fellowship-votes/components/Vote.tsx]: can it be just:
`{active ? 'bg-text-positive shadow-badge-green-background-default' : 'bg-chip-text shadow-secondary-button-background'`
?

File: src/renderer/features/fellowship-votes/components/VotingHistoryList.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    () => performSearch({ records: extendedItems, query, weights: { address: 0.5, name: 1 } }),
+    [items, query],
+  );
+  const { list: deferredItems, isLoading } = useDeferredList({ list: filteredItems, isLoading: !!loading });
+
+  if (!chain) {
+    return null;
+  }
+
+  const shouldRenderLoader = isLoading;
```
Comment (sokolova-an) [inline src/renderer/features/fellowship-votes/components/VotingHistoryList.tsx]: you can make an alias for isLoading the same way as for the list 
`const { list: deferredItems, isLoading: shouldRenderLoader} = useDeferredList({ list: filteredItems, isLoading: !!loading });`

File: src/renderer/features/fellowship-votes/components/VotingHistory.tsx (hunk 11 above)
Comment (Asmadek) [reply src/renderer/features/fellowship-votes/components/VotingHistory.tsx]: It looks like as just ui checking. But it can be discussed 

@johnthecat @sokolova-an

File: src/renderer/shared/ui-entities/RankedAccount/RankedAccount.tsx
Comment (sokolova-an) [inline src/renderer/shared/ui-entities/RankedAccount/RankedAccount.tsx:1]: Can you also add a storybook example for this new component?

File: src/renderer/domains/collectives/model/votes/model.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  $: $list,
+  fulfilled,
+  pending,
+  request,
+} = createDataSource<CollectivesStruct<Record<ReferendumId, Vote[]>>, RequestVotesParams, Vote[]>({
+  initial: {},
+  fn: async ({ api, palletType, referendumId }) => {
+    const votes = await collectivePallet.storage.voting(palletType, api, referendumId);
+
+    return votes.map(mapVote).filter(vote => nonNullable(vote));
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/votes/model.ts]: .filter(nonNullable);

File: src/renderer/features/fellowship-voting-history/model/votes.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+import { fellowshipModel } from './fellowship';
+import { votingHistoryFeatureStatus } from './status';
+
+const gate = createGate<{ referendumId: ReferendumId | null }>({ defaultState: { referendumId: null } });
+
+const $votes = fellowshipModel.$store.map(store => store?.votes ?? []);
+
+const $votesList = combine($votes, gate.state, (votes, { referendumId }) => {
+  if (!votes || referendumId === null) return [];
```
Comment (johnthecat) [inline src/renderer/features/fellowship-voting-history/model/votes.ts]:
```suggestion
  if (nullable(referendumId)) return [];
```

File: src/renderer/domains/collectives/model/members/model.ts [hunk 19]
```diff
@@ -44,15 +44,21 @@ const {
         if (nullable(collectiveMember.member)) continue;
 
         const coreMember = coreMembers.find(x => x.account === collectiveMember.account);
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/members/model.ts]:
```javascript
const coreMember = coreMembers.find(member => member.account === collectiveMember.account);
```

File: src/renderer/domains/collectives/model/members/model.ts [hunk 20]
```diff
@@ -44,15 +44,21 @@ const {
...
-
-        result.push({
-          accountId: collectiveMember.account,
-          rank: collectiveMember.member.rank,
-          isActive: coreMember.status.isActive,
-          lastPromotion: coreMember.status.lastPromotion,
-          lastProof: coreMember.status.lastProof,
-        });
+
+        if (nonNullable(coreMember) && nonNullable(coreMember.status)) {
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/members/model.ts]: i guess this should work
```javascript
if (nonNullable(coreMember?.status)) {
```

File: src/renderer/features/fellowship-voting-history/model/votes.ts [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+import { nonNullable } from '@/shared/lib/utils';
+import { type ReferendumId } from '@/shared/pallet/referenda';
+import { collectiveDomain } from '@/domains/collectives';
+
+import { fellowshipModel } from './fellowship';
+import { votingHistoryFeatureStatus } from './status';
+
+const gate = createGate<{ referendumId: ReferendumId | null }>({ defaultState: { referendumId: null } });
+
+const $votes = fellowshipModel.$store.map(store => store?.votes ?? []);
```
Comment (johnthecat) [inline src/renderer/features/fellowship-voting-history/model/votes.ts]: is it object or array? in `$votesList` you're picking value by referendum id, not index

File: src/renderer/domains/collectives/model/votes/mapper.ts [hunk 22]
```diff
@@ -1,5 +1,16 @@
+import { type ArrayElement } from '@/shared/core';
+import { type collectivePallet } from '@/shared/pallet/collective';
+
 import { type Vote } from './types';
 
-export const mapVote = (): Vote => {
-  return {} as Vote;
+export const mapVote = (
+  vote: ArrayElement<Awaited<ReturnType<typeof collectivePallet.storage.voting>>>,
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/votes/mapper.ts:7]: Damn, son 🥹
Gonna check `ArrayElement`

File: src/renderer/domains/collectives/model/votes/model.ts [hunk 23]
```diff
@@ -24,19 +25,9 @@ const {
...
-        return {
-          accountId: vote.key[1],
-          votes: vote.vote.data,
-          decision: vote.vote.type,
-        };
-      })
-      .filter(vote => nonNullable(vote));
+    const votes = await collectivePallet.storage.voting(palletType, api, referendumId);
+
+    return votes.map(mapVote).filter(vote => nonNullable(vote));
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/votes/model.ts:30]: `votes.map(mapVote).filter(nonNullable);`

File: src/renderer/features/fellowship-voting-history/model/votes.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+const $votesList = combine($votes, gate.state, (votes, { referendumId }) => {
+  if (!votes || referendumId === null) return [];
+
+  return votes[referendumId] ?? [];
+});
+
+const votesUpdate = attachToFeatureInput(votingHistoryFeatureStatus, gate.state);
+
+sample({
+  clock: votesUpdate,
```
Comment (johnthecat) [inline src/renderer/features/fellowship-voting-history/model/votes.ts]: I think it should be just `gate.open`, it's ok because feature works only when view is shown

File: src/renderer/domains/collectives/model/voting/model.ts [hunk 25]
```diff
@@ -47,19 +47,19 @@ const {
 
       if (nullable(accountId) || nullable(referendumId)) continue;
 
-      switch (vote.type) {
+      switch (vote.vote?.type) {
         case 'Aye':
           result.push({
             accountId,
```
Comment (johnthecat) [inline src/renderer/domains/collectives/model/voting/model.ts]: I think now you can map `accountId` and `referendumId` from vote itself, not by index

File: src/renderer/features/fellowship-votes/components/VotingHistory.tsx (hunk 11 above)
Comment (johnthecat) [reply src/renderer/features/fellowship-votes/components/VotingHistory.tsx]: I think it's unnecessary, just checking for additional view state

File: src/renderer/domains/collectives/model/members/service.ts [hunk 26]
```diff
@@ -19,7 +19,12 @@ const findMachingAccount = (wallet: Wallet, member: Member) => {
   });
 };
 
+const isCoreMember = (member: Member | CoreMember) => {
+  return 'isActive' in member && 'lastPromotion' in member && 'lastProof' in member;
+};
+
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/members/service.ts]:
```suggestion
const isCoreMember = (member: Member | CoreMember): boolean => {
      const hasActive = 'isActive' in member;
      const hasPromotion = 'lastPromotion' in member;
      const hasProof = 'lastProof' in member;
  
      return hasActive && hasPromotion && hasProof;
};
```

File: src/renderer/features/fellowship-members/components/Member.tsx [hunk 27]
```diff
@@ -1,16 +1,12 @@
...
-import { AccountExplorers, Address } from '@/shared/ui-entities';
-import { Label } from '@/shared/ui-kit';
-import { type Member as MemberType } from '@/domains/collectives';
+import { RankedAccount } from '@/shared/ui-entities';
+import { type CoreMember } from '@/domains/collectives';
 import { identityModel } from '../model/identity';
 
 type Props = {
-  item: MemberType;
+  item: CoreMember;
```
Comment (tuul-wq) [inline src/renderer/features/fellowship-members/components/Member.tsx:9]: If `item` is always a `CoreMember` maybe it's useful to rename file and component to `CoreMember.tsx`?

File: src/renderer/shared/pallet/collective/storage.ts [hunk 28]
```diff
@@ -107,8 +118,28 @@ export const storage = {
...
    * Votes on a given proposal, if it is ongoing.
    */
-  voting(type: PalletType, api: ApiPromise, keys: (readonly [referendum: ReferendumId, account: AccountId])[]) {
-    return substrateRpcPool.call(() => getQuery(type, api, 'voting').multi(keys)).then(votingResponseSchema.parse);
+  voting(
+    type: PalletType,
+    api: ApiPromise,
+    keys: (readonly [referendum: ReferendumId, account: AccountId])[] | ReferendumId,
+  ) {
+    if (Array.isArray(keys)) {
```
Comment (tuul-wq) [inline src/renderer/shared/pallet/collective/storage.ts]: To reduce nesting and remove redundant `else`
```javascript
if (!Array.isArray(keys)) {
  return substrateRpcPool
    .call(() => getQuery(type, api, 'voting').entries(keys))
    .then(votingWithKeyResponseSchema.parse);
}

const schema = votingResponseSchema.transform(votes =>
  votes.map((vote, index) => ({
    key: {
      referendumId: keys[index]![0],
      accountId: keys[index]![1],
    },
    vote,
  })),
);

return substrateRpcPool.call(() => getQuery(type, api, 'voting').multi(keys)).then(schema.parse);
```

File: src/renderer/shared/core/types/utility.ts [hunk 29]
```diff
@@ -35,3 +35,7 @@ export type Serializable<T> = T extends string | number | null | undefined | boo
 type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };
 
 export type XOR<T, U = object> = T | U extends object ? (Without<T, U> & U) | (Without<U, T> & T) : T | U;
+
+export type ArrayElement<ArrayType extends readonly unknown[]> = ArrayType extends readonly (infer ElementType)[]
```
Comment (tuul-wq) [inline src/renderer/shared/core/types/utility.ts]: Is `readonly` necessary?
Maybe use a shorter version?
```javascript
export type ArrayElement<T extends unknown[]> = T extends (infer E)[] ? E : never;
```

File: src/renderer/features/fellowship-members/components/Member.tsx (hunk 27 above)
Comment (johnthecat) [reply src/renderer/features/fellowship-members/components/Member.tsx:9]: No.

File: src/renderer/features/fellowship-voting-history/model/votes.ts [hunk 30]
```diff
@@ -12,20 +11,23 @@ import { votingHistoryFeatureStatus } from './status';
 
 const gate = createGate<{ referendumId: ReferendumId | null }>({ defaultState: { referendumId: null } });
 
-const $votes = fellowshipModel.$store.map(store => store?.votes ?? []);
-
-const $votesList = combine($votes, gate.state, (votes, { referendumId }) => {
-  if (!votes || referendumId === null) return [];
+const $votesMap = fellowshipModel.$store.map(store => store?.votes ?? []);
```
Comment (johnthecat) [inline src/renderer/features/fellowship-voting-history/model/votes.ts]: It's still an object, not array. Our typescript settings are not strict enough for caching this type of bugs

File: src/renderer/features/fellowship-voting-history/model/votes.ts [hunk 31]
```diff
@@ -12,20 +11,23 @@ import { votingHistoryFeatureStatus } from './status';
...
   return votes[referendumId] ?? [];
 });
 
-const votesUpdate = attachToFeatureInput(votingHistoryFeatureStatus, gate.state);
-
 sample({
-  clock: votesUpdate,
-  filter: ({ data: { referendumId } }) => nonNullable(referendumId),
-  fn: ({ data: { referendumId }, input }) => ({ ...input, referendumId: referendumId! }),
+  clock: gate.state,
```
Comment (johnthecat) [inline src/renderer/features/fellowship-voting-history/model/votes.ts]: let's revert this change as we discussed

**Review by johnthecat — APPROVED** [review]
